File: fotoapparat/src/test/java/io/fotoapparat/parameter/selector/AspectRatioSelectorsTest.java
package io.fotoapparat.parameter.selector;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.ArgumentMatchers;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnitRunner;

import java.util.Collection;

import io.fotoapparat.parameter.Size;

import static io.fotoapparat.test.TestUtils.asSet;
import static java.util.Arrays.asList;
import static junit.framework.Assert.assertEquals;
import static junit.framework.Assert.fail;
import static org.mockito.BDDMockito.given;
import static org.mockito.Mockito.verify;

@RunWith(MockitoJUnitRunner.class)
public class AspectRatioSelectorsTest {

    @Mock
    SelectorFunction<Collection<Size>, Size> sizeSelector;

    @Test
    public void standardRatio() throws Exception {
        // Given
        given(sizeSelector.select(ArgumentMatchers.<Size>anyCollection()))
                .willReturn(new Size(4, 3));

        // When
        Size result = AspectRatioSelectors
                .standardRatio(sizeSelector)
                .select(asList(
                        new Size(4, 3),
                        new Size(8, 6),
                        new Size(10, 10)
                ));

        // Then
        assertEquals(
                new Size(4, 3),
                result
        );

        verify(sizeSelector).select(asSet(
                new Size(4, 3),
                new Size(8, 6)
        ));
    }

    @Test
    public void standardRatioWithTolerance() throws Exception {
        // Given
        given(sizeSelector.select(ArgumentMatchers.<Size>anyCollection()))
                .willReturn(new Size(400, 300));

        // When
        Size result = AspectRatioSelectors
                .standardRatio(sizeSelector, 0.1)
                .select(asList(
                        new Size(400, 300),
                        new Size(410, 300),
                        new Size(800, 600),
                        new Size(790, 600),
                        new Size(100, 100),
                        new Size(160, 90)
                ));

        // Then
        assertEquals(
                new Size(400, 300),
                result
        );

        verify(sizeSelector).select(asSet(
                new Size(400, 300),
                new Size(410, 300),
                new Size(800, 600),
                new Size(790, 600)
        ));
    }

    @Test
    public void wideRatio() throws Exception {
        // Given
        given(sizeSelector.select(ArgumentMatchers.<Size>anyCollection()))
                .willReturn(new Size(16, 9));

        // When
        Size result = AspectRatioSelectors
                .wideRatio(sizeSelector)
                .select(asList(
                        new Size(16, 9),
                        new Size(32, 18),
                        new Size(10, 10)
                ));

        // Then
        assertEquals(
                new Size(16, 9),
                result
        );

        verify(sizeSelector).select(asSet(
                new Size(16, 9),
                new Size(32, 18)
        ));
    }

    @Test
    public void wideRatioWithTolerance() throws Exception {
        // Given
        given(sizeSelector.select(ArgumentMatchers.<Size>anyCollection()))
                .willReturn(new Size(16, 9));

        // When
        Size result = AspectRatioSelectors
                .wideRatio(sizeSelector, 0.1)
                .select(asList(
                        new Size(16, 9),
                        new Size(16, 10),
                        new Size(32, 18),
                        new Size(32, 20),
                        new Size(10, 10)
                ));

        // Then
        assertEquals(
                new Size(16, 9),
                result
        );

        verify(sizeSelector).select(asSet(
                new Size(16, 9),
                new Size(16, 10),
                new Size(32, 18),
                new Size(32, 20)
        ));
    }

    @Test
    public void ratioWithTolerance() throws Exception {
        // Given
        given(sizeSelector.select(ArgumentMatchers.<Size>anyCollection()))
                .willReturn(new Size(110, 100));

        // When
        Size result = AspectRatioSelectors
                .aspectRatio(1.0f, sizeSelector, 0.1)
                .select(asList(
                        new Size(16, 9),
                        new Size(110, 100),
                        new Size(105, 100),
                        new Size(95, 100),
                        new Size(90, 100),
                        new Size(4, 3),
                        new Size(20, 10)
                ));

        // Then
        assertEquals(
                new Size(110, 100),
                result
        );

        verify(sizeSelector).select(asSet(
                new Size(110, 100),
                new Size(105, 100),
                new Size(95, 100),
                new Size(90, 100)
        ));
    }

    @Test
    public void ratioWithNegativeToleranceThrows() throws Exception {
        // Given
        final double NEGATIVE_TOLERANCE = -0.1;
        final double ABOVE_RANGE_TOLERANCE = 1.1;

        try {
            // When
            AspectRatioSelectors
                    .aspectRatio(1.0f, sizeSelector, NEGATIVE_TOLERANCE)
                    .select(asList(
                            new Size(16, 9),
                            new Size(16, 10)));
            fail("Negative aspect ratio tolerance is illegal");
        } catch (IllegalArgumentException ex) {
            // Then exception is throws
        }


        try {
            // When
            AspectRatioSelectors
                    .aspectRatio(1.0f, sizeSelector, ABOVE_RANGE_TOLERANCE)
                    .select(asList(
                            new Size(16, 9),
                            new Size(16, 10)));
            fail("Aspect ratio tolerance >1.0 is illegal");
        } catch (IllegalArgumentException ex) {
            // Then exception is throws
        }
    }
}

File: fotoapparat/src/test/java/io/fotoapparat/selector/AspectRatioSelectorsTest.kt
package io.fotoapparat.selector

import io.fotoapparat.parameter.Resolution
import org.junit.Before
import org.junit.Test
import kotlin.test.assertEquals

class AspectRatioSelectorsTest {

    var receivedResolutions: Iterable<Resolution>? = null

    @Before
    fun setUp() {
        receivedResolutions = null
    }

    @Test
    fun `Filter standard ratios`() {
        // Given
        val sizeSelector: Iterable<Resolution>.() -> Resolution = {
            receivedResolutions = this
            Resolution(4, 3)
        }

        // When
        val result = standardRatio(selector = sizeSelector)(listOf(
                Resolution(4, 3),
                Resolution(8, 6),
                Resolution(10, 10)
        ))

        // Then
        assertEquals(
                expected = Resolution(4, 3),
                actual = result
        )

        assertEquals(
                expected = listOf(
                        Resolution(4, 3),
                        Resolution(8, 6)
                ),
                actual = receivedResolutions
        )
    }

    @Test
    fun `Filter standard ratios with tolerance`() {
        // Given
        val sizeSelector: Iterable<Resolution>.() -> Resolution = {
            receivedResolutions = this
            Resolution(400, 300)
        }

        // When
        val result = standardRatio(
                selector = sizeSelector,
                tolerance = 0.1
        )(listOf(
                Resolution(400, 300),
                Resolution(410, 300),
                Resolution(800, 600),
                Resolution(790, 600),
                Resolution(100, 100),
                Resolution(160, 90)
        ))

        // Then
        assertEquals(
                expected = Resolution(400, 300),
                actual = result
        )

        assertEquals(
                expected = listOf(
                        Resolution(400, 300),
                        Resolution(410, 300),
                        Resolution(800, 600),
                        Resolution(790, 600)
                ),
                actual = receivedResolutions
        )
    }

    @Test
    fun `Filter wide ratios`() {
        // Given
        val sizeSelector: Iterable<Resolution>.() -> Resolution = {
            receivedResolutions = this
            Resolution(16, 9)
        }

        // When
        val result = wideRatio(selector = sizeSelector)(listOf(
                Resolution(16, 9),
                Resolution(32, 18),
                Resolution(10, 10)
        ))

        // Then
        assertEquals(
                expected = Resolution(16, 9),
                actual = result
        )

        assertEquals(
                expected = listOf(
                        Resolution(16, 9),
                        Resolution(32, 18)
                ),
                actual = receivedResolutions
        )
    }

    @Test
    fun `Filter wide ratios with tolerance`() {
        // Given
        val sizeSelector: Iterable<Resolution>.() -> Resolution = {
            receivedResolutions = this
            Resolution(16, 9)
        }

        // When
        val result = wideRatio(
                selector = sizeSelector,
                tolerance = 0.1
        )(listOf(
                Resolution(16, 9),
                Resolution(16, 10),
                Resolution(32, 18),
                Resolution(32, 20),
                Resolution(10, 10)
        ))

        // Then
        assertEquals(
                expected = Resolution(16, 9),
                actual = result
        )

        assertEquals(
                expected = listOf(
                        Resolution(16, 9),
                        Resolution(16, 10),
                        Resolution(32, 18),
                        Resolution(32, 20)
                ),
                actual = receivedResolutions
        )
    }

    @Test
    fun `Filter custom aspect ratio with tolerance`() {
        // Given
        val sizeSelector: Iterable<Resolution>.() -> Resolution = {
            receivedResolutions = this
            Resolution(110, 100)
        }

        // When
        val result = aspectRatio(
                aspectRatio = 1.0f,
                selector = sizeSelector,
                tolerance = 0.1
        )(listOf(
                Resolution(16, 9),
                Resolution(110, 100),
                Resolution(105, 100),
                Resolution(95, 100),
                Resolution(90, 100),
                Resolution(4, 3),
                Resolution(20, 10)
        ))

        // Then
        assertEquals(
                expected = Resolution(110, 100),
                actual = result
        )

        assertEquals(
                expected = listOf(
                        Resolution(110, 100),
                        Resolution(105, 100),
                        Resolution(95, 100),
                        Resolution(90, 100)
                ),
                actual = receivedResolutions
        )
    }

    @Test(expected = IllegalArgumentException::class)
    fun `Custom Ratio with tolerance less than 0 should throw exception`() {
        // Given
        val sizeSelector: Iterable<Resolution>.() -> Resolution = {
            Resolution(16, 9)
        }

        // When
        aspectRatio(
                aspectRatio = 1.0f,
                selector = sizeSelector,
                tolerance = -0.1
        )(listOf(
                Resolution(16, 9),
                Resolution(16, 10)
        ))

        // Then exception is thrown
    }

    @Test(expected = IllegalArgumentException::class)
    fun `Custom ratio with tolerance above 1 should throw exception`() {
        // Given
        val sizeSelector: Iterable<Resolution>.() -> Resolution = {
            Resolution(16, 9)
        }

        // When
        aspectRatio(
                aspectRatio = 1.0f,
                selector = sizeSelector,
                tolerance = 1.1
        )(listOf(
                Resolution(16, 9),
                Resolution(16, 10)
        ))

        // Then exception is thrown
    }
}
